File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        s = 0;
        e = len(nums)
        mid = (s+e)//2

        while s < e:
            
            if nums[mid] == target :
                return mid

            if nums[s] <= nums[mid]:
                if nums[s] <= target and nums[mid] >= target:
                    e = mid
                else:
                    s = mid +1
            else:
                if target >= nums[mid] and target < nums[s]:
                    s = mid +1
                else:
                    e = mid
            
            mid = (s+e)//2
            #print(f"s: {s}, nums[s]: {nums[s]}, mid: {mid}, nums[mid]: {nums[mid]},  e: {e}, nums[e]: {nums[e]}")
        return -1

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        int last = nums.size()-1;
        if (last == 0)
        {
            return nums[0];
        }
        int s = 0;
        int e = nums.size()-1;
        int mid = (s+e)/2;
        while(s <= mid && mid < e)
        {
            cout << "s: " << s << " mid: " << mid << " e: " << e << '\n';
            if(mid < last && nums[mid] > nums[mid+1])
            {
                return nums[mid+1];
            }
            else if(mid < last && nums[mid] < nums[mid+1])
            {
                if (nums[mid] > nums[last])
                {
                    s = mid;
                }
                else 
                {
                    e = mid;
                }
                mid = (s+e)/2;
            }
        }
        if (nums[mid] > nums[e])
        {
            return nums[e];
        }
        if (nums[mid] < nums[s])
        {
            return nums[mid];
        }
        return nums[0];
    }
};
